### PR TITLE
EVG-14845 Correctly determine blocked status for tasks

### PR DESCRIPTION
**Pull request by khelif96**
[EVG-14845](https://jira.mongodb.org/browse/EVG-14845)

### Description 
We weren't able to properly determine if a task was blocked or not because of some weirdness with how embedded documents in arrays are handled in aggregations. This fixes that and properly  queries the `depends_on.unattainable` field in a task and determines if it is blocked or not.
It does this by adding a stage before the check that saves the unattainable state of a task in a temporary field.

### CODE REVIEW

File: model/task/task.go
Comment (khelif96) [inline model/task/task.go:1385]: removed this check because it would query the db for every single dependent task. Instead we should just rely on the unattainable field in the task since that should represent if a task is blocked or not.

**Comment by bsamek** [comment]
I've removed myself and assigned @ybrill , since he wrote the dependent tasks logic so is more likely to catch problems here.

File: model/task/db.go
Comment (bsamek) [inline model/task/db.go:146]: One note (but need not block this PR, since I've removed myself): It might be better to use a `$project` for the `$filter` instead of an `$addFields`. If you're never going to use the array of subdocuments again, then there's no need to keep them in the document. Since they can be pretty big, it might be helpful to reduce them to an array of size 0 or 1 immediately instead of keeping them around.

Put another way, your current approach is
1. Add a new field, an array of size either 0 or 1
2. Add a new field based on the size of the new field in #1

Instead, you could
1. Change the current field to an array of size 0 or 1
2. Change the field again to a boolean based on the size of the array

The first approach uses 2 `$addFields`. The latter uses 2 `$project`s.

File: model/task/task.go
Comment (ablack12) [inline model/task/task.go:2809]: is this a mistake?

File: model/task/db.go
Comment (ablack12) [inline model/task/db.go:232]: Is this the only place we use HasUnattainableDeps? In which case we don't need to add it to the task document, since everything is done within the aggregation. You can just use a local global variable string or something.

**Review by ablack12 — CHANGES_REQUESTED** [review]

**Comment by ybrill** [comment]
a different strategy could be to adjust [the checks for not blocked](https://github.com/evergreen-ci/evergreen/blob/4c1b1fbcffd1ad79de7c96ecf1a4e65417309793/model/task/db.go#L220) and blocked to reduce the array to a single true/false value before comparing with true/false.

Something like:
```
  {
    "case": {"$ne": [{"$reduce": 
      {
        input: "$depends_on",
        initialValue: false,
        in: { "$or": ["$$value.unattainable", "$$this.unattainable"] }
      }}, true]},
    "then": "Will Run",
  }
```

File: model/task/db.go
Comment (khelif96) [reply model/task/db.go]: Good call

File: model/task/task.go
Comment (khelif96) [reply model/task/task.go]: nice catch

**Comment by khelif96** [comment]
> a different strategy could be to adjust the existing checks [for not blocked](https://github.com/evergreen-ci/evergreen/blob/4c1b1fbcffd1ad79de7c96ecf1a4e65417309793/model/task/db.go#L220) and [blocked](https://github.com/evergreen-ci/evergreen/blob/4c1b1fbcffd1ad79de7c96ecf1a4e65417309793/model/task/db.go#L241) to reduce the array to a single true/false value before comparing with true/false.
> 
> Something like:
> 
> ```
>   {
>     "case": {"$ne": [{"$reduce": 
>       {
>         input: "$depends_on",
>         initialValue: false,
>         in: { "$or": ["$$value.unattainable", "$$this.unattainable"] }
>       }}, true]},
>     "then": "Will Run",
>   }
> ```

This is magic

File: model/task/db.go
Comment (ybrill) [inline model/task/db.go:224]: would it make sense to order these three cases:
1. `TaskUnscheduled`
2. `TaskStatusBlocked`
3. `TaskWillRun`

This would help that we wouldn't need to check `isAttainable` and `ActivatedKey` for this case.

File: model/task/db.go
Comment (ablack12) [inline model/task/db.go:141]: could you drop a comment explaining what this is doing, just for future reference?

**Review by ablack12 — COMMENTED** [review]
Jonathan has a clearer understanding of the aggregation haha so LGTM once he's good

File: model/task/db.go
Comment (ybrill) [reply model/task/db.go:224]: If this works, maybe we could get away with using the reduction in place of the comparison to `isAttainable`. This way we wouldn't even need to `addFields`. Something like
```
	"case": bson.M{
		"$and": []bson.M{
			{"$eq": []string{"$" + StatusKey, evergreen.TaskUndispatched}},
			{"$eq": []interface{}{ []interface{}{bson.M{"$reduce": bson.M{
			    "input":        "$" + DependsOnKey,
			    "initialValue": false,
			    "in":           bson.M{"$or": []interface{}{"$$" + bsonutil.GetDottedKeyName("value", DependencyUnattainableKey), "$$" + bsonutil.GetDottedKeyName("this", DependencyUnattainableKey)}},
		        },
		        }, true}}},
		},
	},
	"then": evergreen.TaskStatusBlocked,
```

Comment (khelif96) [reply model/task/db.go:224]: That does make sense. But i think it might be better to be explicit about the requirements for each status as it helps with readability and makes it easier to understand what the code is doing. What do you think?

File: model/task/db.go
Comment (ybrill) [inline model/task/db.go:246]: This check seems superfluous since we've already established the task isn't blocked in the previous case.

**Review by ybrill — APPROVED** [review]
> What do you think?

That makes sense to me.

**Review by ybrill — APPROVED** [review]

**Pull request merged**
